feat: Added initial support for dialog, path and fs plugins

emirror-de opened the pull request:
This PR introduces the initial support for `dialog`, `path` and `fs` plugins.

This is not yet 100% complete and tested, but at least an initial implementation to build on.

bicarlsen commented on src/fs.js:
```diff
@@ -0,0 +1,1105 @@
+function transformCallback(callback, once = false) {
+    return window.__TAURI_INTERNALS__.transformCallback(callback, once);
+}
+class Channel {
```
Why are you including the whole `Channel` class here instead of calling `__TAURI__.core.Channel` where needed?

emirror-de replied:
I removed the class definition and updated the calls to `Channel` as you suggested in c11448e

bicarlsen commented on src/path.rs:
````diff
@@ -0,0 +1,790 @@
+/// # }
+/// ```
+#[inline(always)]
+pub async fn resolve_resource(resource_path: &str) -> crate::Result<PathBuf> {
````
Should this accept `AsRef<Path>` instead of `&str`?

emirror-de replied:
Updated in [7b12c94a](https://github.com/emirror-de/tauri-sys/commit/7b12c94a40bdba63c6c604ae905b9ee018f7cfd2)

bicarlsen commented on src/path.rs:
````diff
@@ -0,0 +1,790 @@
+/// # }
+/// ```
+#[inline(always)]
+pub async fn resolve(paths: impl IntoIterator<Item = &str>) -> crate::Result<PathBuf> {
````
Should this accept `AsRef<Path>` instead of `&str`?

emirror-de replied:
Updated in [a66b598a](https://github.com/emirror-de/tauri-sys/commit/a66b598a7a456b301290f8de5a5b37f5ab108813)

bicarlsen commented on src/path.rs:
````diff
@@ -0,0 +1,790 @@
+/// # }
+/// ```
+#[inline(always)]
+pub async fn join(paths: impl IntoIterator<Item = &str>) -> crate::Result<PathBuf> {
````
Should `paths` accept `AsRef<Path>` instead of `&str`?

emirror-de replied:
Updated in ad71f9e

bicarlsen commented on src/path.rs:
````diff
@@ -0,0 +1,790 @@
+/// # }
+/// ```
+#[inline(always)]
+pub async fn dirname(path: &str) -> crate::Result<PathBuf> {
````
Should this accept `AsRef<Path>` instead of `&str`?

emirror-de replied:
Updated in 1fb776d

bicarlsen commented on src/path.rs:
````diff
@@ -0,0 +1,790 @@
+/// # }
+/// ```
+#[inline(always)]
+pub async fn extname(path: &str) -> crate::Result<PathBuf> {
````
I don't know if it is necessary to implement this as [`Path::extension`](https://doc.rust-lang.org/std/path/struct.Path.html#method.extension) exists.

bicarlsen commented on src/path.rs:
````diff
@@ -0,0 +1,790 @@
+/// # }
+/// ```
+#[inline(always)]
+pub async fn basename(path: &str, ext: Option<&str>) -> crate::Result<PathBuf> {
````
I don't know if it is necessary to implement this as [`Path::file_name`](https://doc.rust-lang.org/std/path/struct.Path.html#method.file_name) exists.

emirror-de replied:
Removed in 792a75b

bicarlsen commented on src/path.rs:
````diff
@@ -0,0 +1,790 @@
+/// # }
+/// ```
+#[inline(always)]
+pub async fn is_absolute(path: &str) -> crate::Result<bool> {
````
I don't know if it is necessary to implement this as [`Path::is_absolute`](https://doc.rust-lang.org/std/path/struct.Path.html#method.is_absolute) exists.

emirror-de replied:
Removed in e2b49dd

bicarlsen reviewed (commented):
Thanks! I had some questions about some implementation details. A few occurred multiple times, so I copy-pasted the comment to mark each spot I noticed it.

emirror-de commented on src/fs.rs:
```diff
@@ -0,0 +1,512 @@
+
+    let raw = inner::readDir(
+        path,
+        serde_wasm_bindgen::to_value(&FsDirOptions {
```
Updated in https://github.com/JonasKruckenberg/tauri-sys/pull/68/commits/77bc8d581a853355c1e083fcd6f7a4447e5b0d92

emirror-de commented on src/fs.rs:
```diff
@@ -0,0 +1,512 @@
+}
+
+#[derive(Deserialize, Clone, PartialEq, Debug)]
+pub struct FileEntry {
```
Yes it should. I updated it in [5cf028de](https://github.com/emirror-de/tauri-sys/commit/5cf028deffa93edb8798d9f296aaec18c012c4d2)

emirror-de commented on src/fs.rs:
````diff
@@ -0,0 +1,512 @@
+/// ```
+///
+/// Requires [`allowlist > fs > readDir`](https://tauri.app/v1/api/js/fs) to be enabled.
+pub async fn read_dir_all(path: &Path, dir: BaseDirectory) -> crate::Result<Vec<FileEntry>> {
````
As far as I can see, there is no `readDirAll` function in v2? I only can find [readDir](https://v2.tauri.app/reference/javascript/fs/#readdir)

emirror-de commented:
Yes sure, I will also re-request a review as soon as I finished walking through.

emirror-de commented on src/fs.rs:
```diff
@@ -0,0 +1,512 @@
+        #[wasm_bindgen(catch)]
+        pub async fn readDir(dir: &str, options: JsValue) -> Result<JsValue, JsValue>;
+        #[wasm_bindgen(catch)]
+        pub async fn removeDir(dir: &str, options: JsValue) -> Result<(), JsValue>;
```
Yes, removed `remove_*` functions of v1 and added `remove` function. [320ffd82](https://github.com/emirror-de/tauri-sys/commit/320ffd828cec3d3ad87a22d8be26020b3d53e6c0)

emirror-de commented on src/fs.rs:
```diff
@@ -0,0 +1,512 @@
+        #[wasm_bindgen(catch)]
+        pub async fn removeFile(source: &str, options: JsValue) -> Result<(), JsValue>;
+        #[wasm_bindgen(catch)]
+        pub async fn renameFile(
```
Updated in [80b2fe5c](https://github.com/emirror-de/tauri-sys/commit/80b2fe5c7c8d6556d99a2111114d500e6dd3a391).

emirror-de commented on src/fs.rs:
````diff
@@ -0,0 +1,512 @@
+/// ```
+///
+/// Requires [`allowlist > fs > renameFile`](https://tauri.app/v1/api/js/fs) to be enabled.
+pub async fn rename_file(
````
`rename_file` has been removed, `rename` is now for directories and files.

emirror-de commented:
@bicarlsen thanks for the review. I must admit, the initial PR state was just a copy-paste of `v1` version with collected JS and adjusted to work for my urgent requirements. But I wanted to share and show that there has been some work started in case anyone is happy to work on it as long as I am not able to.

Despite to this, I adjusted most of your suggestions. Unfinished threads are still left open.

I think there is much more work to do on this, as I did not update the documentation yet to match the ones from the reference.

Some ideas I had during my walk through the code:

- In `path.rs` should we replace the functions that return base directories and add a generic one that takes `BaseDirectory` as param? This would reduce the amount of code on Rust side quite a bit.
- I now sometimes added the `options` parameter as second argument to the Rust functions as for example in [fs::remove](https://github.com/JonasKruckenberg/tauri-sys/blob/e2b49dd79451dd5fda65c9d8675075c89cea2321/src/fs.rs#L324). Should we adjust the rest as well to get closer to the JS API? Or decide it by the amount of parameters in the specific functions?

bicarlsen commented:
@emirror-de, thanks for all the work :)

### 1
For the `BaseDirectory` functions, I leave it to you to choose whatever feels more ergonomic. I don't use these functions, so don't have a sense. The code for the individual functions is simple enough I'm not worried about maintaining the extra lines.

### 2
In general I would prefer deviating from the JS API if it means we can be more Rusty.
e.g.
```js
function myFunction(requiredArg, optionalArg)
```
should go to
```rust
fn my_function(required_arg)
fn my_function_with_options(required_arg, optional_arg)
```
